File: bindings/swift/examples/CameraRecord.swift
// CameraRecord.swift
//
// ASPIRATIONAL EXAMPLE: no `Mediaway` Swift package exists yet — no
// `mediaway-device-ffi` / `mediaway-encoder-ffi` C ABI crates have shipped
// either. This shows the target ergonomics for a future Swift binding over
// Mediaway's C ABI: a C bridging header wrapped in idiomatic Swift (classes
// with `deinit`/explicit `close()`, `throws`/`try` instead of raw C status
// codes). See ../README.md and docs/spec/c-ffi.md.
//
// Same shape as ../ScreenRecord.swift, with a camera source in place of a
// screen source: open camera + microphone capture, build an H.264 auto
// encoder sized to the camera's negotiated geometry, run the same
// platform-agnostic `record` loop that polls both sources and writes
// synthetic grey NV12 placeholder frames into the encode session, then
// finish the session and write the fragmented MP4 bytes to `out_camera.mp4`.
//
// `CameraCapture`/`Microphone` (concrete backends) and the `VideoCapture`/
// `AudioCapture` protocols they conform to are all part of the imagined
// `Mediaway` package — same as `AutoVideoEncoder`/`EncodeSession`/`VideoFrame`
// in ../EncodeToMp4.swift. Only `record(...)` and the top-level script below
// are example/app code.
//
// Run (once the real package exists):
//     swift run CameraRecord

import Foundation
import Mediaway

// MARK: - Platform-agnostic record loop

/// Poll `video` and `audio` until `duration` elapses, writing one synthetic
/// grey NV12 placeholder frame into `session` per captured video frame, and
/// draining (but not yet doing anything with) audio frames.
///
/// Typed entirely against the `VideoCapture`/`AudioCapture` protocols — this
/// is the exact same function used in ../ScreenRecord.swift; it does not
/// know or care whether the concrete backend behind `video` is a camera or a
/// screen. `audio` is `nil` when the caller could not open a microphone;
/// recording continues video-only in that case.
func record(
    video: VideoCapture,
    audio: AudioCapture?,
    session: EncodeSession,
    width: Int,
    height: Int,
    duration: TimeInterval
) {
    let deadline = Date().addingTimeInterval(duration)
    // Synthetic NV12 placeholder (Y=128, UV=128 → grey), same shape as
    // ../EncodeToMp4.swift: width*height Y bytes + width*height/2
    // interleaved UV bytes.
    let ySize = width * height
    let uvSize = width * height / 2
    let greyNV12 = Data(repeating: 128, count: ySize + uvSize)

    var pts: Int64 = 0
    while Date() < deadline {
        do {
            if try video.pollFrame() != nil {
                // A real backend would convert the captured pixels (often
                // GPU-resident camera memory) to NV12 here; this example
                // writes the synthetic placeholder instead and releases the
                // frame back to the OS once it is no longer needed.
                video.releaseFrame()

                let frame = VideoFrame(
                    pts: pts,
                    duration: 1,
                    width: width,
                    height: height,
                    pixelFormat: .nv12,
                    data: greyNV12
                )
                try session.writeFrame(frame)
                pts += 1
            }
        } catch {
            print("camera_record: capture error (\(error))")
            break
        }

        // Audio track not wired in this example — frames are drained only.
        guard let audio else { continue }
        while (try? audio.pollFrame()) != nil {}
    }
}

// MARK: - Entry point

let fps = 30
let frameRate = Rational(1, fps)

// ── 1. Open camera capture (device 0 = default/first camera) — throws if
//    that camera is not available. Handle gracefully instead of crashing. ──
let camera: CameraCapture
do {
    camera = try CameraCapture(device: 0, frameRate: frameRate)
} catch {
    print("camera_record: camera unavailable (\(error))")
    exit(0)
}
print("camera_record: \(camera.width)x\(camera.height) camera stream")

// ── 2. Open microphone capture — also fallible; recording continues
//    video-only if no microphone is available. ─────────────────────────────
var mic: Microphone?
do {
    mic = try Microphone(sampleRate: Rational(1, 48_000))
    print("camera_record: microphone ready")
} catch {
    print("camera_record: microphone unavailable (\(error)) — continuing without audio")
    mic = nil
}

do {
    // ── 3. Build the encoder config at the camera's negotiated geometry,
    //    then open the auto encoder + encode session. ────────────────────────
    var config = VideoEncodeConfig.defaults(
        codec: .h264,
        width: camera.width,
        height: camera.height,
        frameRate: frameRate
    )
    config.bitrateBps = 4_000_000

    let encoder = try AutoVideoEncoder(config: config)
    let session = try EncodeSession(encoder: encoder)

    // ── 4. Run the shared record loop for 3 seconds. ─────────────────────────
    record(
        video: camera,
        audio: mic,
        session: session,
        width: camera.width,
        height: camera.height,
        duration: 3
    )

    // ── 5. Close both capture objects. ───────────────────────────────────────
    camera.close()
    mic?.close()

    // ── 6. Finish the encode session and write the fragmented MP4 bytes. ────
    let mp4Bytes = try session.finish()
    try mp4Bytes.write(to: URL(fileURLWithPath: "out_camera.mp4"))

    print("camera_record: -> out_camera.mp4 (\(mp4Bytes.count) bytes)")
} catch {
    camera.close()
    mic?.close()
    print("camera_record: recording failed (\(error))")
    exit(1)
}

File: bindings/kotlin/examples/CameraRecord.kt
/*
 * Camera + mic capture -> encode -> fragmented MP4 — aspirational quick-start example.
 *
 * ASPIRATIONAL EXAMPLE: no `mediaway` Kotlin/Java package exists yet. This file
 * shows the target ergonomics for a future Kotlin binding over Mediaway's C ABI
 * (JNI under the hood, wrapped idiomatically: Closeable resources used with
 * `.use { }`, Kotlin exceptions instead of raw error codes). The API is also
 * meant to read naturally from plain Java, so no Kotlin-only surface (inline
 * value classes, etc.) is used in the public shape. See ../README.md and
 * docs/spec/c-ffi.md.
 *
 * Same shape as screen capture would be (see e.g. ../csharp/ScreenRecord.cs),
 * with a camera source instead of a screen source: open a camera capture + a
 * microphone (both fallible — the specific device may not be available), build
 * an "auto video encode" config at the capture's real geometry, and reuse the
 * exact same building blocks as EncodeToMp4.kt (AutoVideoEncoder ->
 * EncodeSession -> writeFrame -> finish), plus one small platform-agnostic
 * `record()` function that glues capture to encode. `record()` is typed purely
 * against the `VideoCapture` / `AudioCapture` interfaces, so it has no idea a
 * camera (rather than a screen) is underneath — the exact same function would
 * work unchanged for screen capture.
 *
 * Run (once the real package exists):
 *     kotlinc CameraRecord.kt -include-runtime -d camera-record.jar
 *     java -jar camera-record.jar
 */

import io.mediaway.container.Rational
import io.mediaway.device.AudioCapture
import io.mediaway.device.CameraCapture
import io.mediaway.device.CaptureUnavailableException
import io.mediaway.device.Microphone
import io.mediaway.device.VideoCapture
import io.mediaway.encoder.AutoVideoEncodeConfig
import io.mediaway.encoder.AutoVideoEncoder
import io.mediaway.encoder.EncodeSession
import io.mediaway.encoder.NoBackendException
import io.mediaway.encoder.VideoFrame
import java.io.File
import java.time.Duration
import java.time.Instant

private const val FPS = 30
private const val SECONDS = 3L
private const val AUDIO_SAMPLE_RATE = 48_000
private const val BITRATE_BPS = 4_000_000L

/**
 * Poll [video] (and drain [audio], if present) until [duration] elapses.
 *
 * Writes a synthetic grey NV12 placeholder frame into [session] for every
 * polled video frame — this example never touches the real captured pixels.
 * Typed purely against the [VideoCapture] / [AudioCapture] interfaces: this
 * function has no idea which concrete source (camera, screen, ...) it was
 * handed, so the exact same function works unchanged for screen capture.
 */
private fun record(
    video: VideoCapture,
    audio: AudioCapture?,
    session: EncodeSession,
    duration: Duration,
) {
    val deadline = Instant.now().plus(duration)

    // Synthetic NV12 placeholder (stand-in for real captured pixels): grey
    // Y=128, UV=128. Layout is width*height Y bytes followed by
    // width*height/2 interleaved UV bytes.
    val width = video.width
    val height = video.height
    val greyNv12 = ByteArray(width * height + width * height / 2) { 128.toByte() }

    var pts = 0L
    while (Instant.now().isBefore(deadline)) {
        val frame = video.pollFrame()
        if (frame != null) {
            // The polled frame may reference GPU-resident memory that this
            // example never touches — it must still be released back to the
            // OS once we're done with it, even though we encode a synthetic
            // buffer instead.
            video.releaseFrame()

            session.writeFrame(
                VideoFrame(
                    pts = pts++,
                    duration = 1,
                    width = width,
                    height = height,
                    pixelFormat = VideoFrame.PixelFormat.NV12,
                    data = greyNv12,
                )
            )
        }

        // Drain polled audio frames — not wired into an audio track yet.
        while (audio?.pollFrame() != null) {
            // no-op: draining only
        }
    }
}

fun main() {
    // -- 1. Open capture sources. Both are fallible: the specific camera or
    //       microphone may not be available on this machine. ---------------
    val videoCapture: VideoCapture = try {
        CameraCapture.open(deviceIndex = 0, frameRate = Rational(1, FPS))
    } catch (e: CaptureUnavailableException) {
        println("camera_record: camera unavailable (${e.message}) -- exiting")
        return
    }

    val audioCapture: AudioCapture? = try {
        Microphone.open(sampleRate = Rational(1, AUDIO_SAMPLE_RATE))
    } catch (e: CaptureUnavailableException) {
        // Recording without audio is a valid degraded mode -- log and continue.
        println("camera_record: microphone unavailable (${e.message}) -- continuing without audio")
        null
    }

    // The capture settles on its own stream geometry (whatever the camera
    // actually negotiated) -- read it back here rather than assuming it.
    val width = videoCapture.width
    val height = videoCapture.height
    println("camera_record: ${width}x$height camera" + if (audioCapture != null) ", mic ready" else "")

    // -- 2. Open the encoder + encode session (same building blocks as
    //       EncodeToMp4.kt), sized to the capture's real geometry. ----------
    val config = AutoVideoEncodeConfig
        .h264Defaults(width = width, height = height, frameRateNum = FPS, frameRateDen = 1)
        .withBitrateBps(BITRATE_BPS)

    val encoder = try {
        AutoVideoEncoder.open(config)
    } catch (e: NoBackendException) {
        println("camera_record: no H.264 encoder backend available on this platform (${e.message})")
        videoCapture.close()
        audioCapture?.close()
        return
    }

    // -- 3. Run the platform-agnostic record loop, then finish encoding and
    //       write the result. ------------------------------------------------
    val mp4Bytes = encoder.use { enc ->
        EncodeSession(enc).use { session ->
            record(videoCapture, audioCapture, session, Duration.ofSeconds(SECONDS))

            videoCapture.close()
            audioCapture?.close()

            session.finish()
        }
    }

    val outFile = File("out_camera.mp4")
    outFile.writeBytes(mp4Bytes)

    println("camera_record: ${width}x$height -> ${outFile.name} (${mp4Bytes.size} bytes)")
}
